src/stateReducer.js

```diff
@@ -0,0 +1,124 @@
+const ACTIONS = {
+  ADD_NEW_CART_ITEM: "add-new-cart-item",
+  BUY_NOW_HANDLER: "buy-now-handler",
+  UPDATE_WISHLIST: "update-wishlist",
+  INCREMENT_CART_COUNT: "increment-cart-count",
+  DECREMENT_CART_COUNT: "decrement-cart-count",
+  REMOVE_FROM_CART: "remove-from-cart",
+  REMOVE_FROM_WISHLIST: "remove-from-wishlist",
+  SORT: "Sort",
+  FILTER: "filter"
+};
+
+export const reducer = (state, action) => {
+  switch (action.type) {
+    case ACTIONS.ADD_NEW_CART_ITEM: {
+      let flag = false;
+      const newCartList = state.cartList.map((cartItem) => {
+        if (cartItem.id === action.payload.id) {
+          flag = true;
+          return {
+            ...cartItem,
+            quantity: cartItem.quantity + action.payload.quantity
+          };
+        } else {
+          return cartItem;
+        }
+      });
+
+      if (flag) {
+        return { ...state, cartList: newCartList };
+      } else {
+        return {
+          ...state,
+          cartList: [
+            ...state.cartList,
+            { ...action.payload.item, quantity: action.payload.quantity }
+          ]
+        };
+      }
+    }
+
+    case ACTIONS.BUY_NOW_HANDLER: {
+      const findItem = state.cartList.find(
+        (item) => item.id === action.payload.id
+      );
+      if (findItem) {
+        return state;
+      } else {
+        return {
+          ...state,
+          cartList: [
+            ...state.cartList,
+            { ...action.payload.item, quantity: action.payload.quantity }
+          ]
+        };
+      }
+    }
+    case ACTIONS.UPDATE_WISHLIST: {
+      let flag = false;
+      state.wishList.map((item) => {
+        if (item.id === action.payload.id) flag = true;
+      });
+      if (flag) return state;
+      else
+        return {
+          ...state,
+          wishList: [...state.wishList, { ...action.payload.item, quantity: 1 }]
```

Comment:
Same here, I feel this part of code is a bit confusing, to make it readable using find method to first find the item and based on it is present or not, return or concat payload